### PR TITLE
[NAE-1577] Task reffed change behavior does not propagate - correction

### PR DESCRIPTION
**Pull request by renczesstefan**
# Description

Fixes issue when a field's behaviour is changed in task reference and this change did not apply. It is fixed via requesting reload on the referenced task.

Fixes: [NAE-1577]

## Dependencies

No new dependencies was used, imported or implemented.

### Third party dependencies

No new third party dependencies was used, imported or implemented.

### Blocking Pull requests

There are no dependencies on other PR.

## How Has Been This Tested?

Tested it using the provided Petri Net files in Jira issue [NAE-1577]. I imported the Petri Nets into the application engine, and using example app and followed the reproduction steps that where described in Jira issue.

### Test Configuration

Host OS is macOS Monterey 12.1, the app was running on Angular 10.2.3 with Node 12.22.9 and NPM 6.14.15.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes have been checked, personally or remotely, with @machacjozef
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have resolved all conflicts with the target branch of the PR
- [x] I have updated and synced my code with the target branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes:
    - [x] Lint test
    - [x] Unit tests
    - [x] Integration tests
- [x] I have checked my contribution with code analysis tools:
    - [x] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [x] [Snyk](https://app.snyk.io/org/netgrif)
- [x] I have made corresponding changes to the documentation:
    - [x] Developer documentation
    - [x] User Guides
    - [x] Migration Guides

[NAE-1577]: https://netgrif.atlassian.net/browse/NAE-1577?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NAE-1577]: https://netgrif.atlassian.net/browse/NAE-1577?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts [hunk 1]
```diff
@@ -296,6 +297,25 @@ export abstract class TaskContentService implements OnDestroy {
...
     }
 
+    private updateReferencedField(chFields: ChangedFields, field: DataField<any>, frontendActions: Change): void {
+        if (this._fieldConverterService.resolveType(field) === FieldTypeResource.TASK_REF) {
+            this._taskDataReloadRequest$.next(frontendActions ? frontendActions : undefined);
+            return;
+        }
+        const updatedField = chFields[field.stringId];
+        Object.keys(updatedField).forEach(key => {
+            switch (key) {
```
Comment (minop) [inline projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts:307]: this switch feels very familiar. Is it not a duplicate of some other change resolution code elsewhere?

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts [hunk 2]
```diff
@@ -314,6 +334,24 @@ export abstract class TaskContentService implements OnDestroy {
...
+            .some(taskId => taskId !== this.task.stringId && this.taskFieldsIndex[taskId].fields[changedField]);
+    }
+
+    private getReferencedTransitionId(changedField: string): string {
+        let taskFieldsIndexId = '';
+        if (!!this.taskFieldsIndex) {
+            taskFieldsIndexId = Object.keys(this.taskFieldsIndex).find(taskId =>
+                Object.keys(this.taskFieldsIndex[taskId].fields).includes(changedField));
+        }
+        return this.taskFieldsIndex[taskFieldsIndexId].transitionId;
```
Comment (minop) [inline projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts:346]: if `this.taskFieldsIndex` is falsty it still gets accessed with the `[]` operator and will almost certainly throw an error.

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts [hunk 3]
```diff
@@ -314,6 +334,24 @@ export abstract class TaskContentService implements OnDestroy {
...
+    }
+
+    private getReferencedTaskId(changedField: string): string {
+        let taskFieldsIndexId = '';
```
Comment (minop) [inline projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts:350]: this variable is redundant

**Review by minop — CHANGES_REQUESTED** [review]

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts (hunk 1 above)
Comment (renczesstefan) [reply projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts:309]: This contains different logic than the similar switch in the same file.

**Review by minop — DISMISSED** [review]

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts (hunk 3 above)
Comment (renczesstefan) [reply projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts]: Corrected

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts (hunk 2 above)
Comment (renczesstefan) [reply projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts]: Corrected

File: projects/netgrif-components-core/src/lib/task-content/model/task-fields.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import {DataField} from '../../data-fields/models/abstract-data-field';
+
+export interface TaskFields {
```
Comment (mladoniczky) [inline projects/netgrif-components-core/src/lib/task-content/model/task-fields.ts:3]: This interface is now exported in public-api file for task-content module.
Please include it in exported objects.
https://github.com/netgrif/components/blob/NAE-1577_2/projects/netgrif-components-core/src/lib/task-content/public-api.ts

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts [hunk 5]
```diff
@@ -226,10 +225,13 @@ export abstract class TaskContentService implements OnDestroy {
...
-            if (!!this.taskFieldsIndex[chFields.taskId] && !!this.taskFieldsIndex[chFields.taskId][changedField]) {
-                this.updateField(chFields, this.taskFieldsIndex[chFields.taskId][changedField], frontendActions);
+            if (!!this.taskFieldsIndex[chFields.taskId] && !!this.taskFieldsIndex[chFields.taskId].fields
+                && !!this.taskFieldsIndex[chFields.taskId].fields[changedField]) {
+                this.updateField(chFields, this.taskFieldsIndex[chFields.taskId].fields[changedField], frontendActions);
             } else if (this.isFieldInTaskRef(changedField)) {
-                this._taskDataReloadRequest$.next(frontendActions ? frontendActions : undefined);
+                const taskId = this.getReferencedTaskId(changedField);
+                if (!!taskId)
+                    this.updateReferencedField(chFields, this.taskFieldsIndex[taskId].fields[changedField], frontendActions);
```
Comment (mladoniczky) [inline projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts:234]: `this.taskFieldsIndex[taskId].fields` is not checked if the property exists as on line 228.

**Review by mladoniczky — CHANGES_REQUESTED** [review]

File: projects/netgrif-components-core/src/lib/task-content/model/task-fields.ts (hunk 4 above)
Comment (renczesstefan) [reply projects/netgrif-components-core/src/lib/task-content/model/task-fields.ts:3]: Corrected

File: projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts (hunk 5 above)
Comment (renczesstefan) [reply projects/netgrif-components-core/src/lib/task-content/services/task-content.service.ts:235]: Corrected

**Review by mladoniczky — APPROVED** [review]

**Review by minop — APPROVED** [review]

**Pull request merged**